### DISCUSSION TITLE
How to get enum value

### DISCUSSION THREAD

**Discussion by marcotrumpet** [vfsfitvnm/frida-il2cpp-bridge#259] [Q&A]
Hi, I need to access an enum in order to pass it to a method.

That's the enum using `Il2Cpp.dump()` 
<img width="396" alt="Screenshot 2023-03-07 alle 12 11 22" src="https://user-images.githubusercontent.com/35219231/223406049-b73924fe-2f0a-4c93-b804-7cb8638e7f19.png">

And that's the same using an external software
<img width="496" alt="Screenshot 2023-03-07 alle 12 12 26" src="https://user-images.githubusercontent.com/35219231/223406271-f2f475cb-1dd1-41a6-9948-874ade4e5b35.png">

My problem is that if I try to access it with 
```ts
const AssemblyCSharp = Il2Cpp.Domain.assembly('Assembly-CSharp').image;
const ScoreType = AssemblyCSharp.class('com.spaceape.config.ScoreType');
```
I get an error
`il2cpp: couldn't find class com.spaceape.config.ScoreType in Assembly-CSharp.dll, did you mean com.spaceape.unityui.SafeRegions?`

Do you have any idea? Could it be the assembly the wrong one?

Thanks in advance

**Comment by vfsfitvnm** [comment]
That class doesn't reside in `Assembly-CSharp`, but  `RakshaModel`

**Reply by marcotrumpet** [reply]
Oh nice thanks it worked, now I need to figure out how to hook the method passing this argument. Right now I have
```ts
    const AssemblyCSharp = Il2Cpp.Domain.assembly('Assembly-CSharp').image;
    const TapInteraction = AssemblyCSharp.class('BeatStar.RhythmGame.TapInteraction');

    const ScoreType = Il2Cpp.Domain.assembly('RakshaModel').image.class('com.spaceape.config.ScoreType');

    var method = TapInteraction.method('SetTapScore');

    method.implementation = (tapScore: number, scoreType: Il2Cpp.Object) => {
        // I assume need to change here scoreType with the new one
        // but the following doesn't work
       tapScore = 500;
       scoreType.field<Il2Cpp.ValueType>(scoreType.toString()).value = ScoreType.field<Il2Cpp.ValueType>('APLUS').value;
    };
```

**Reply by vfsfitvnm** [reply]
It doesn't work because the JS arguments aren't bound with the native ones.
I'm not sure what's the logic in there, but I think this is enough:
```ts
    TapInteraction.method("SetTapScore").implementation = function(tapScore: number, scoreType: Il2Cpp.Object) {
      return this.method("SetTapScore").invoke(500, scoreType);
    };
```

**Reply by marcotrumpet** [reply]
Thank you so much! (Sorry if questions seems dumb but it's my first try to this package)

**Reply by ChuJiani** [reply]
Before starting a new one, it would be a good idea to check if there have been any previous discussions on the same topic☺️

**Reply by vfsfitvnm** [reply]
It's ok tought, a proper documentation/walkthrought is currently lacking - and it may be challenging for rookies to get started - I can't really blame them. I'd rather answer to trivial questions instead of having no interactions because they can't solve a trivial problem so they abandon the library

**Reply by marcotrumpet** [reply]
To be honest, I checked previous discussions about enums. My implementation was ok (because of the previous discussion) but I was searching in Assembly-CSharp instead of RakshaModel.
By the way if you have plans to have a complete wiki and I can help in some ways (even if I'm noob) let me know, happy to give something back to this amazing project.

In the end I was searching in the wrong spot but ended up doing what I needed to do with Il2Cpp module https://github.com/marcotrumpet/beatstar-perfect-plus

**Reply by ChuJiani** [reply]
> To be honest, I checked previous discussions about enums. My implementation was ok (because of the previous discussion) but I was searching in Assembly-CSharp instead of RakshaModel. By the way if you have plans to have a complete wiki and I can help in some ways (even if I'm noob) let me know, happy to give something back to this amazing project.

It would be helpful to provide some best practices (in the wiki) for this library, and newcomers can start by imitating. The existing examples in the wiki are not very comprehensive, and many novices are not familiar with the syntax of typescript. If some complete examples can be given, I think some AI models such as chatgpt or copilot can be used to assist in understanding and development. When I first started a year ago, as a novice in il2cpp and typescript, it was really torturous to start only relying on the wiki

**Reply by marcotrumpet** [reply]
If someone more experienced than me can provide a table of contents or some sort of "ways to tackle documentation" (I don't know where to start from), I can try to write something.. anyways thanks for the patience..

**Reply by vfsfitvnm** [reply]
Yeah - the wiki is indeed not much helpful.
The author (me) already knows how to use his library, and pretending to be a novice is hard - so I don't really know what and how I should explain things to novices! 
(I should probably iterate the issues and base the new one on them instead).

Any help/idea is appreciated :)

**Reply by ChuJiani** [reply]
> If someone more experienced than me can provide a table of contents or some sort of "ways to tackle documentation" (I don't know where to start from), I can try to write something.. anyways thanks for the patience..

Maybe you can refer to the [documentation](https://vuejs.org/guide/introduction.html) of the vue framework for ideas. 

To better illustrate and verify the usage of the library, it would be useful to create a simple, original game based on il2cpp as an example(to avoid any potential copyright or other issues).

Recently, I've had a lot going on and haven't had as much time to focus on this😭. If I have more availability later on, I might be able to lend a hand.

**Reply by vfsfitvnm** [reply]
> To better illustrate and verify the usage of the library, it would be useful to create a simple, original game based on il2cpp as an example(to avoid any potential copyright or other issues).

Yeah, that's exactly where the problem is. I have no idea how to create a game in Unity and honestly I'm too lazy to learn - maybe we can use an open source project with a compatible license. I've been using this [template](https://learn.unity.com/project/2d-platformer-template) so far, but it's way too minimal for building a decent guide on top of it.
Ideally, we should also have a cross-platform and cross-unity-version test suite too.

> Recently, I've had a lot going on and haven't had as much time to focus on this :sob:. If I have more availability later on, I might be able to lend a hand.

Thanks for your willingness to help! Unfortunately I don't have much free time, too :/

**Reply by vfsfitvnm** [reply]
https://github.com/vfsfitvnm/frida-il2cpp-bridge/issues/261

**Comment by SaberMansori** [comment]
how i can get enum like this ~>

`
public enum myType
{
	one
	two,
	three,
	four
}
`
 
thank you.

**Reply by ekologic** [reply]
Quick search gives this [Get enum value](https://github.com/vfsfitvnm/frida-il2cpp-bridge/discussions/356)
